### PR TITLE
Implement initial tuple syntax

### PR DESCRIPTION
**Pull request by joshuawarner32**
Step 1 of N toward implementing #4465

Note, this is a long way from complete - but it should start to unblock other work.

### CODE REVIEW

**Comment by joshuawarner32** [comment]
Ooooh CI caught an interesting problem!

Example (False.roc, line 438):
```
            Task.fromResult
                (
                    a, b <- binaryOp ctx
                    if a == b then
                        -1
                    else
                        0
                )
```

Part of this change was moving from parenthesized-expressions allowing multi-backpassing, to not allowing that (matching what we do in both lists and record literals).

This is because there's a fundamental ambiguity between that meaning (a) a parenthesized expr with multi-backpassing, or (b) a tuple of the value `a` and a single-backpassing closure.

I'm currently treating a parenthesized expression as a degenerate case of a tuple - but we could instead 

Technically we could keep them separate and first check if this parsed as a parenthesized expression, and then later try parsing as a tuple - but that's both inefficient since we're re-parsing, and rather awkward from a teaching perspective.

IMO it's already not immediately obvious what the above syntax example is supposed to do; I think this would be a reasonable re-write?

```
            Task.fromResult
                binaryOp ctx \a, b ->
                    if a == b then
                        -1
                    else
                        0
```

That honestly seems even easier to mentally parse than the current code in False.roc.

@rtfeldman, thoughts?

**Comment by ayazhafiz** [comment]
My 2c: I think the reasonable re-write is okay, and another possible option is that expressions in tuples match be indented at least as deep as where they start, i.e.

```
            Task.fromResult
                (
                    a, b <- binaryOp ctx
                         if a == b then
                            -1
                         else
                             0
                )
```

could parse as a tuple but otherwise you backtrace and parse as a backpacking. Although I don't think this resolves an ambiguity in the following situation
```
            Task.fromResult
                (
                    a,
                    b <- binaryOp ctx
                    if a == b then
                        -1
                    else
                        0
                )
```

so imo making this always attempt to parse as a tuple seems reasonable

File: crates/reporting/src/error/parse.rs [hunk 1]
```diff
@@ -1007,6 +1007,26 @@ fn to_expr_in_parens_report<'a>(
...
+        EInParens::Empty(pos) => {
+            let surroundings = Region::new(start, pos);
+            let region = LineColumnRegion::from_pos(lines.convert_pos(pos));
+
+            let doc = alloc.stack([
+                alloc.reflow("I am partway through parsing a parenthesized expression or tuple:"),
+                alloc.region_with_subregion(lines.convert_region(surroundings), region),
+                alloc.concat([
+                    alloc.reflow(r"I was expecting to see an expression next."),
+                    alloc.reflow(r"Note, Roc doesn't use '()' as a null type."),
```
Comment (rtfeldman) [inline crates/reporting/src/error/parse.rs:1019]: Nice note! 👍

**Review by rtfeldman — APPROVED** [review]
Looks great! 💯

Excited to see this happening! 😃

**Pull request merged**